scipy/optimize/_linprog.py

```diff
@@ -135,742 +141,1009 @@ def linprog_terse_callback(xk, **kwargs):
 
-    if phase == 1:
-        m = T.shape[0]-2
-    elif phase == 2:
-        m = T.shape[0]-1
-    else:
-        raise ValueError("Argument 'phase' to _solve_simplex must be 1 or 2")
-
-    if phase == 2:
-        # Check if any artificial variables are still in the basis.
-        # If yes, check if any coefficients from this row and a column
-        # corresponding to one of the non-artificial variable is non-zero.
-        # If found, pivot at this term. If not, start phase 2.
-        # Do this for all artificial variables in the basis.
-        # Ref: "An Introduction to Linear Programming and Game Theory"
-        # by Paul R. Thie, Gerard E. Keough, 3rd Ed,
-        # Chapter 3.7 Redundant Systems (pag 102)
-        for pivrow in [row for row in range(basis.size)
-                       if basis[row] > T.shape[1] - 2]:
-            non_zero_row = [col for col in range(T.shape[1] - 1)
-                            if abs(T[pivrow, col]) > tol]
-            if len(non_zero_row) > 0:
-                pivcol = non_zero_row[0]
-                _apply_pivot(T, basis, pivrow, pivcol)
-                nit += 1
-
-    if len(basis[:m]) == 0:
-        solution = np.zeros(T.shape[1] - 1, dtype=np.float64)
+    References
+    ----------
+    .. [2] Andersen, Erling D. "Finding all linearly dependent rows in
+           large-scale linear programming." Optimization Methods and Software
+           6.3 (1995): 219-227.
+    .. [5] Andersen, Erling D., and Knud D. Andersen. "Presolving in linear
+       programming." Mathematical Programming 71.2 (1995): 221-245.
+
+    """
+    # ideas from Reference [5] by Andersen and Andersen
+    # however, unlike the reference, this is performed before converting
+    # problem to standard form
+    # There are a few advantages:
+    #  * artificial variables have not been added, so matrices are smaller
+    #  * bounds have not been converted to constraints yet. (It is better to
+    #    do that after presolve because presolve may adjust the simple bounds.)
+    # There are many improvements that can be made, namely:
+    #  * implement remaining checks from [5]
+    #  * loop presolve until no additional changes are made
+    #  * implement additional efficiency improvements in redundancy removal [2]
+
+    undo = []               # record of variables eliminated from problem
+    # constant term in cost function may be added if variables are eliminated
+    c0 = 0
+    complete = False        # complete is True if detected infeasible/unbounded
+    x = np.zeros(c.shape)   # this is solution vector if completed in presolve
+
+    status = 0              # all OK unless determined otherwise
+    message = ""
+
+    # Standard form for bounds (from _clean_inputs) is list of tuples
+    # but numpy array is more convenient here
+    # In retrospect, numpy array should have been the standard
+    bounds = np.array(bounds)
+    lb = bounds[:, 0]
+    ub = bounds[:, 1]
+    lb[np.equal(lb, None)] = -np.inf
```

Comment:
By easier, do you mean more efficient? I didn't know that vectorized code on object arrays would be less efficient than a pure Python solution, and since it's far from a bottleneck I didn't profile it.
We can consider this for a style cleanup later, but I really think we should avoid style changes to existing code in this PR, which is for unifying simplex and interior point pre- and post-processing.